sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java

```diff
@@ -170,77 +170,135 @@ public int readInteger() {
    *  }
    */
   public void readBatch(
-      int total,
-      int offset,
+      ParquetReadState state,
       WritableColumnVector values,
-      int maxDefinitionLevel,
       VectorizedValuesReader valueReader,
       ParquetVectorUpdater updater) throws IOException {
-    int left = total;
-    while (left > 0) {
+    int offset = state.offset;
+    long rowId = state.rowId;
+
+    while (state.hasMoreInPage(offset, rowId)) {
       if (this.currentCount == 0) this.readNextGroup();
-      int n = Math.min(left, this.currentCount);
-      switch (mode) {
-        case RLE:
-          if (currentValue == maxDefinitionLevel) {
-            updater.updateBatch(n, offset, values, valueReader);
-          } else {
-            values.putNulls(offset, n);
-          }
-          break;
-        case PACKED:
-          for (int i = 0; i < n; ++i) {
-            if (currentBuffer[currentBufferIdx++] == maxDefinitionLevel) {
-              updater.update(offset + i, values, valueReader);
+      int n = Math.min(state.valuesToReadInBatch + state.offset - offset, this.currentCount);
```

Comment:
`n` is the number of values to read?

Isn't it just `Math.min(state.valuesToReadInBatch - offset, this.currentCount)`?

Why needs to add `state.offset`?

Reply:
because `state.valuesToReadInBatch` is the number of values left in the batch. Suppose the initial batch size is 1000, and `valuesToReadInBatch` is 400. This means we've read 600 values and so `state.offset` and `offset` are both 600 at the beginning. We'll need to use `offset - state.offset` to get how many values we've read so far.